### DISCUSSION TITLE
Delete (Remove) of highlighted annotation from PDF

### DISCUSSION THREAD

**Discussion by Carnegie-eng** [pymupdf/PyMuPDF#2825] [Q&A] [answered]
Hi, I am tryin to delete annotations (highlight marker) from PDF document. But the following code does not work.
Could you correct the code for me ? Thank you so much for your support !

```
doc = fitz.open(path)
page = doc[number]

for annot in page.annots():
  annot.delete_responses()

doc.save(newPath)
```

**Comment by JorjMcKie** [comment, marked as answer]
Your code would only delete the responses to an annotation. These are things like PopUp annotation and any thread of text annotations marked as "IRT" ("in response to"). Your intention seems to be the following:
```python
annot = page.first_annot
while annot:
    if annot.type[0] == fitz.PDF_ANNOT_HGIGHLIGHT:
        annot = page.delete_annot(annot)
    else:
        annot = annot.next
```

**Comment by Carnegie-eng** [comment]
Thank you so much for quick support !

I changed the typo to 
`if annot.type[0] == fitz.PDF_ANNOT_HIGHLIGHT:`
and the expected result was obtained !

Thank you.

**Comment by occludedpixel** [comment]
If anyone gets the following error when trying to delete annotations:

```
>>> for annot in doc[2].annots():
...     doc[2].delete_annot(annot)
...
Traceback (most recent call last):
  File "<stdin>", line 2, in <module>
  File "/Volumes/dan_t7/poetry_cache/virtualenvs/pdf-J9RHzUSY-py3.11/lib/python3.11/site-packages/fitz/__init__.py", line 297, in __repr__
    return "'%s' annotation on %s" % (self.type[1], str(parent))
                                                    ^^^^^^^^^^^
ReferenceError: weakly-referenced object no longer exists
```

I believe this is related to [#2484](https://github.com/pymupdf/PyMuPDF/discussions/2484#discussioncomment-6228878) (point 2) and also in the [documentation appendix](https://pymupdf.readthedocs.io/en/latest/app3.html#ensuring-consistency-of-important-objects-in-pymupdf) which makes it necessary to make the assignment `page = doc[number]` then calling `page.delete_annot(annot)`.

**Reply by JorjMcKie** [reply]
You gave the correct answer yourself! Whenever you do `doc[i].any_method()` the page gets created, the method executed, the page deleted.
Any page-dependent object created by `.any_method()` will immediately be orphaned afterwards: it will have not parent page in the moment of being born!

**Reply by JorjMcKie** [reply]
Apart from that:
It is always bad coding style and a high-risk approach to change the basis of an iterator while iterating over it!
To delete all annotations, there exist two clean iteration approaches:
```python
annot = page.first_annot
while annot:
    annot = page.delete_annot(annot)  # the method delivers None or the next annotation
```
Or like this:
```python
for xref, atype in page.annot_xrefs():  # delivers xrefs of annots, fields and links
    try:
        annot = page.load_annot(xref)
        page.delete_annot(annot)
    except Exception:
        pass  # this xref was not annotation
```
